[12.0][ADD] base_search_custom_field_filter: Add custom filters for fields via UI

pedrobaeza opened the pull request:
This module allows to define custom filters in the search views for an specific field belonging to the document or any other related document.

This nature makes the definition quite technical, but once done, it adds the element in the UI for regular user use.

Configuration
=============

* Go to *Settings > Technical > User Interface > Custom Field Filters*.
* Create a new record, and define following information:

  * The **Model** for which you are defining the filter. It will appear in all
    the search views of this model.
  * The label you want to see on the search line on the **Name** field. This
    field allows translations for proper UI in different languages.
  * The **Expression**, which is the field chain string with dot notation.
    Examples: `product_id`, `product_id.seller_ids.name`, `partner_id.lang`.
  * Optionally, you can fill **Position After** for indicating after which
    existing field (technical name) the filter will appear. If empty or not
    found, the filter will be added at the end.
* You can reorder records for determining sorting for multiple filters for the
  same model with the arrow handle in the left part.

Usage
=====

* Go to the menu entry for which you have defined the custom field filter.
* On the search bar, type anything.
* In the filter list, you will see the line for the element you have defined.

As demo data, a custom field filter is included for sample purposes.

![ir_ui_custom_field_filter](https://user-images.githubusercontent.com/7165771/78916432-bf7f7480-7a8d-11ea-94e9-de7b105fb177.png)

Steps for trying this sample:

* Install `contacts` module.
* Go to *Contacts*.
* Type "english" and you'll find the filter "Language" at the end.

![contact_search](https://user-images.githubusercontent.com/7165771/78916440-c6a68280-7a8d-11ea-85ac-ce76878ade3c.png)

cc @Tecnativa TT23306

pedrobaeza committed:
[ADD] base_search_custom_field_filter: Add custom filters for fields via UI

This module allows to define custom filters in the search views for an specific
field belonging to the document or any other related document.

This nature makes the definition quite technical, but once done, it adds the
element in the UI for regular user use.

Configuration
=============

* Go to *Settings > Technical > User Interface > Custom Field Filters*.
* Create a new record, and define following information:

  * The **Model** for which you are defining the filter. It will appear in all
    the search views of this model.
  * The label you want to see on the search line on the **Name** field. This
    field allows translations for proper UI in different languages.
  * The **Expression**, which is the field chain string with dot notation.
    Examples: `product_id`, `product_id.seller_ids.name`, `partner_id.lang`.
  * Optionally, you can fill **Position After** for indicating after which
    existing field (technical name) the filter will appear. If empty or not
    found, the filter will be added at the end.
* You can reorder records for determining sorting for multiple filters for the
  same model with the arrow handle in the left part.

Usage
=====

* Go to the menu entry for which you have defined the custom field filter.
* On the search bar, type anything.
* In the filter list, you will see the line for the element you have defined.

As demo data, a custom field filter is included for sample purposes.

Steps for trying this sample:

* Install `contacts` module.
* Go to *Contacts*.
* Type "english" and you'll find the filter "Language" at the end.

carlosdauden commented on base_search_custom_field_filter/models/base.py:
```diff
@@ -0,0 +1,59 @@
+                    'name': 'ir_ui_custom_filter_%s' % custom_filter.id,
+                    'string': custom_filter.name,
+                    'filter_domain': (
+                        "[('%s', '=', 'self')]" % custom_filter.expression),
```
Remove apostrophes to don't search "self" string literally
```suggestion
                        "[('%s', '=', self)]" % custom_filter.expression),
```

carlosdauden replied:
Changed here: https://github.com/Tecnativa/server-ux/pull/1

pedrobaeza replied:
Thanks, typo error while transcribing the result of the investigation.

simahawk commented on base_search_custom_field_filter/models/ir_ui_custom_field_filter.py:
```diff
@@ -0,0 +1,42 @@
+    name = fields.Char(required=True, translate=True)
+    expression = fields.Char(required=True)
+    position_after = fields.Char(
+        help="Optional field name for putting the filter after that one. "
```
```suggestion
        help="Optional field name for putting the filter after another. "
```

pedrobaeza replied:
Please read it again. The sentence is saying exactly the purpose.

simahawk commented on base_search_custom_field_filter/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,5 @@
+This module allows to define custom filters in the search views for an specific
+field belonging to the document or any other related document.
+
+This nature makes the definition quite technical, but once done, it adds the
```
you can probably drop this last sentence

pedrobaeza replied:
Why? I want to highlight that. Remember this is general "functional user" README, and a warning about this is always advisable.

simahawk reviewed (approved):
looks nice, tnx :)

sergio-teruel commented on base_search_custom_field_filter/models/base.py:
```diff
@@ -0,0 +1,59 @@
+        for custom_filter in custom_filters:
+            field = custom_filter._get_related_field()
+            res['ir_ui_custom_filter_%s' % custom_filter.id] = (
+                field.get_description(self.env))
```
You can not add a "non-existent" field in fields get, this crash write models due to mail_thread model...
https://github.com/odoo/odoo/blob/cc6eb096f61d84f35585005a9f891f62adbb28e8/addons/mail/models/mail_thread.py#L553
If you add this field in this method crash becouse not exists in record
https://github.com/odoo/odoo/blob/cc6eb096f61d84f35585005a9f891f62adbb28e8/addons/mail/models/mail_thread.py#L317

pedrobaeza replied:
Thanks for checking. Please try now with the latest push.

sergio-teruel replied:
Ok. Fixed

sergio-teruel commented on base_search_custom_field_filter/models/base.py:
```diff
@@ -0,0 +1,67 @@
+                    'name': 'ir_ui_custom_filter_%s' % custom_filter.id,
+                    'string': custom_filter.name,
+                    'filter_domain': (
+                        "[('%s', '=', self)]" % custom_filter.expression),
```
Why is equal?? Can not be a like operator?

pedrobaeza replied:
Yeah, good question, probably it should be `ilike`. Let me check.

pedrobaeza replied:
Fixed other way without side effects.

chienandalu reviewed (approved):
Tested on runbot. Really cool! :+1:

carlosdauden committed:
[FIX] base_search_custom_field_filter: Avoid problem with records with same field content

If the field you are looking for is a relational one, and contains the same value
in different records, when you unfold the results and select one of them, the results
of the search belongs to the first ocurrence instead the specific selected record.

This is due to Odoo JS code using same operator in this kind of search when filter_domain
is used in the filter. We avoid this letting the standard field code to act, but
injecting a custom attr that overwrites the left leaf of the domain.

pedrobaeza commented:
@Tardo, can you please check 280311f55d60898a6a4092089fbbf2333e3debb5?

Tardo commented on base_search_custom_field_filter/static/src/js/search_input.js:
```diff
@@ -0,0 +1,22 @@
+    ManyToOneField.include({
+        make_domain: function (name, operator, facetValue) {
+            name = this.attrs.custom_field_filter || name;
+            return this._super(name, operator, facetValue);
```
To avoid linter warnings....
```suggestion
            var name_n = this.attrs.custom_field_filter || name;
            return this._super(name_n, operator, facetValue);
```

Tardo commented on base_search_custom_field_filter/static/src/js/search_input.js:
```diff
@@ -0,0 +1,22 @@
+    search_inputs.Field.include({
+        make_domain: function (name, operator, facetValue) {
+            name = this.attrs.custom_field_filter || name;
+            return this._super(name, operator, facetValue);
```
To avoid linter warnings...
```suggestion
            var name_n = this.attrs.custom_field_filter || name;
            return this._super(name_n, operator, facetValue);
```

Tardo reviewed (approved):
Good work! only a few non-blocking comments

pedrobaeza commented:
@Tardo changes done

pedrobaeza commented:
/ocabot merge
